ui/app/components/auth/form/saml.ts

```diff
@@ -44,136 +42,104 @@ export default class AuthFormSaml extends AuthBase {
-    // start watching the popup window and the current one
-    this.watchPopup.perform(samlWindow);
-    this.watchCurrent.perform(samlWindow);
-
-    let resp;
-    try {
-      resp = await this.pollForToken(samlWindow, { path });
-      this.closeWindow(samlWindow);
-    } catch (error) {
-      this.cancelLogin(samlWindow, errorMessage(error));
-      return;
-    }
+  cancelLogin() {
+    this.login.cancelAll();
+  }
 
-    // We've got a response from the polling request
-    // pass MFA data or use the Vault token (client_token) to continue the auth
-    const mfa_requirement = resp?.mfa_requirement;
-    const client_token = resp?.client_token;
-    if (mfa_requirement) {
-      this.handleMfa(mfa_requirement, path);
-      return;
-    }
-    if (client_token) {
-      this.continueLogin({ token: client_token });
-      return;
-    }
+  // Fetch role to get sso_service_url which is where popup is opened
+  async fetchSamlRole({ namespace = '', path = '', role = '' }): Promise<SamlRole> {
+    // Create the client verifier and challenge
+    const verifier = uuid();
+    const clientChallenge = await this.generateClientChallenge(verifier);
+    const acsUrl = this.generateAcsUrl(path, namespace);
+    const clientType = SamlWriteSsoServiceUrlRequestClientTypeEnum.BROWSER; // 'browser'
+    // Kick off the authentication flow by generating the SSO service URL
+    // It requires the client challenge generated from the verifier. We'll
+    // later provide the verifier to match up with the challenge on the server
+    // when we poll for the Vault token by its returned token poll ID.
+    const { data } = <SamlSsoServiceUrlApiResponse>(
+      await this.api.auth.samlWriteSsoServiceUrl(path, { acsUrl, clientChallenge, clientType, role })
+    );
+    return {
+      ...data,
+      clientVerifier: verifier,
+    };
+  }
 
-    // If there's a problem with the SAML exchange the auth workflow should fail earlier.
-    // Including this catch just in case, though it's unlikely this will be hit.
-    this.handleSAMLError('Missing token. Please try again.');
-    return;
-  });
+  async startSAMLAuth(ssoServiceUrl: string) {
+    const win = window;
+    const POPUP_WIDTH = 500;
+    const POPUP_HEIGHT = 600;
+    const left = win.screen.width / 2 - POPUP_WIDTH / 2;
+    const top = win.screen.height / 2 - POPUP_HEIGHT / 2;
+    const samlWindow = win.open(
+      ssoServiceUrl,
+      'vaultSAMLWindow',
+      `width=${POPUP_WIDTH},height=${POPUP_HEIGHT},resizable,scrollbars=yes,top=${top},left=${left}`
+    );
+
+    return samlWindow;
+  }
 
-  async pollForToken(samlWindow: Window, { path = '' }) {
-    // Poll every one second for the token to become available
+  async exchangeSAMLTokenPollID(fetchedRole: SamlRole, { path = '' }) {
     const WAIT_TIME = Ember.testing ? 50 : 1000;
-    const MAX_TIME = 180; // 3 minutes in seconds
+    const MAX_TRIES = Ember.testing ? 3 : 180; // 180 is 3 minutes in seconds
```

Comment:
I think this may have been why some of the saml tests took longer, even in testing we were still polling 180 times.